File: src/main/java/com/intera/util/web/servlet/filter/JsonpCallbackFilter.java
package com.intera.util.web.servlet.filter;

import java.io.IOException;
import java.io.OutputStream;
import java.util.Map;

import javax.servlet.Filter;
import javax.servlet.FilterChain;
import javax.servlet.FilterConfig;
import javax.servlet.ServletException;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class JsonpCallbackFilter implements Filter {

    private static final Logger log = LoggerFactory.getLogger(JsonpCallbackFilter.class);

    public void init(FilterConfig fConfig) throws ServletException {}

    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
        HttpServletRequest httpRequest = (HttpServletRequest) request;
        HttpServletResponse httpResponse = (HttpServletResponse) response;

        @SuppressWarnings("unchecked")
		Map<String, String[]> parms = httpRequest.getParameterMap();

        String callback = null;
        if(parms.containsKey("callback") && null != (callback = parms.get("callback")[0])) {
            if(log.isDebugEnabled())
                log.debug("Wrapping response with JSONP callback '" + callback + "'");
            OutputStream out = httpResponse.getOutputStream();
            out.write(new String(callback + "(").getBytes());
            chain.doFilter(request, httpResponse);
            out.write(new String(");").getBytes());
            
            httpResponse.setContentType("text/javascript;charset=UTF-8");
        } else {
            chain.doFilter(request, response);
        }
    }

    public void destroy() {}
}

File: src/main/java/com/intera/util/web/servlet/filter/JsonpCallbackFilter.scala
package com.intera.util.web.servlet.filter

import java.io.{ByteArrayOutputStream, OutputStream}
import java.util.Map
import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
import javax.servlet.{Filter, FilterChain, FilterConfig, ServletRequest, ServletResponse}

import com.typesafe.scalalogging.slf4j.{StrictLogging, LazyLogging}

import scala.collection.JavaConverters._


class JsonpCallbackFilter extends Filter with StrictLogging {
  def init(fConfig: FilterConfig) {
  }

  def doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain) {
    val httpRequest: HttpServletRequest = request.asInstanceOf[HttpServletRequest]
    val httpResponse: HttpServletResponse = response.asInstanceOf[HttpServletResponse]

    val params: Map[String, Array[String]] = httpRequest.getParameterMap.asScala.toMap.asInstanceOf[Map[String, Array[String]]];

    if (params.containsKey("callback")) {
      logger.debug("Wrapping response with JSONP callback '" + params.get("callback")(0) + "'")
      val out: OutputStream = httpResponse.getOutputStream
      val wrapper: GenericResponseWrapper = new GenericResponseWrapper(httpResponse)
      chain.doFilter(request, wrapper)
      val outputStream: ByteArrayOutputStream = new ByteArrayOutputStream
      outputStream.write(new String(params.get("callback")(0) + "(").getBytes)
      outputStream.write(wrapper.getData)
      outputStream.write(new String(");").getBytes)
      val jsonpResponse: Array[Byte] = outputStream.toByteArray
      wrapper.setContentType("text/javascript;charset=UTF-8")
      wrapper.setContentLength(jsonpResponse.length)
      out.write(jsonpResponse)
      out.close
    }
    else {
      chain.doFilter(request, response)
    }
  }

  def destroy {
  }

}